src/accelerate/utils/dataclasses.py

```diff
@@ -1594,26 +1611,53 @@ def __post_init__(self):
+        if self.fsdp_version is None:
+            self.fsdp_version = int(os.environ.get(env_prefix + "VERSION", "1"))
+
+        if self.reshard_after_forward is None:
+            self.reshard_after_forward = os.environ.get(
+                env_prefix + "RESHARD_AFTER_FORWARD",
+                True if self.fsdp_version == 2 else "FULL_SHARD",
+            )
+        if isinstance(self.reshard_after_forward, str):
+            if self.fsdp_version == 2:
+                self.reshard_after_forward = str_to_bool(self.reshard_after_forward.lower())
             else:
-                self.sharding_strategy = ShardingStrategy[self.sharding_strategy.upper()]
+                # We need to remap based on custom enum values for user readability
+                if self.reshard_after_forward.upper() in FSDP_SHARDING_STRATEGY:
+                    self.reshard_after_forward = FSDP_SHARDING_STRATEGY.index(self.reshard_after_forward.upper()) + 1
+                if isinstance(self.reshard_after_forward, int) or self.reshard_after_forward.isdigit():
+                    self.reshard_after_forward = ShardingStrategy(int(self.reshard_after_forward))
+                else:
+                    self.reshard_after_forward = ShardingStrategy[self.reshard_after_forward.upper()]
+        if self.fsdp_version != 2 and isinstance(self.reshard_after_forward, bool):
+            raise ValueError(
+                "reshard_after_forward set to bool. This is not supported in FSDP1, please set to a `str` or an instance of `torch.distributed.fsdp.fully_sharded_data_parallel.ShardingStrategy`"
+            )
 
         if self.cpu_offload is None:
             self.cpu_offload = str_to_bool(os.environ.get(env_prefix + "OFFLOAD_PARAMS", "False")) == 1
         if isinstance(self.cpu_offload, bool):
-            self.cpu_offload = CPUOffload(offload_params=self.cpu_offload)
+            if self.fsdp_version == 2:
+                if not self.cpu_offload:
+                    warnings.warn("Offload_params is set to False, however FSDP2 always offloads parameters and runs optimizer step on CPU. This will be overridden to True.")
+                self.cpu_offload = CPUOffloadPolicy()  # TODO: enable pin memory config
+            else:
+                self.cpu_offload = CPUOffload(offload_params=self.cpu_offload)
+        if self.fsdp_version != 2 and isinstance(self.cpu_offload, CPUOffloadPolicy):
+            raise ValueError(
+                "cpu_offload set to `torch.distributed.fsdp.CPUOffloadPolicy`. This is not supported in FSDP1, please set to a `bool` or an instance of `torch.distributed.fsdp.CPUOffload`"
+            )
+        if self.fsdp_version == 2 and not isinstance(self.cpu_offload, CPUOffloadPolicy):
+            raise ValueError(
+                "cpu_offload set to `bool` or `torch.distributed.fsdp.CPUOffload`. This is not supported in FSDP2, please set to an instance of `torch.distributed.fsdp.CPUOffloadPolicy`"
+            )
```

Comment:
I feel like we can simplify this

```suggestion
    if isinstance(self.cpu_offload, CPUOffloadPolicy):
        err = "`cpu_offload` set to `torch.distributed.fsdp.CPUOffloadPolicy`."
        if self.fsdp_version != 2:
            raise ValueError(f"{err} This is not supported in FSDP1, please set to a `bool` or an instance of `torch.distributed.fsdp.CPUOffload`")
        else:
            raise ValueError(f"{err} This is not supported in FSDP2, please set to an instance of `torch.distributed.fsdp.CPUOffloadPolicy`")
```